src/System/Process/Typed.hs

```diff
@@ -621,14 +622,17 @@ startProcess pConfig'@ProcessConfig {..} = liftIO $ do
         ec <-
           if multiThreadedRuntime
             then P.waitForProcess pHandle
-            else
-              let loop = do
-                    threadDelay 20000
+            else do
+              switchTime <- (fromIntegral . (`div` 1000) . ctxtSwitchTime)
+                        <$> getConcFlags
+              let loop delay = do
+                    let delay' = min switchTime (delay * 2)
+                    threadDelay delay'
                     mec <- P.getProcessExitCode pHandle
                     case mec of
-                      Nothing -> loop
+                      Nothing -> loop delay'
                       Just ec -> pure ec
-               in loop
+              loop 1000
```

Comment:
This will start with `1 ms` unconditionally, which means no more than 1000 process calls per second on non-threaded - very strong throttling.

I recommend to actually start with `1`.